libraries/chain/sps_evaluator.cpp

```diff
@@ -0,0 +1,157 @@
+
+#include <steem/chain/database.hpp>
+
+#include <steem/chain/steem_evaluator.hpp>
+#include <steem/chain/sps_objects.hpp>
+#include <steem/chain/sps_helper.hpp>
+
+
+namespace steem { namespace chain {
+
+using steem::chain::create_proposal_evaluator;
+
+void create_proposal_evaluator::do_apply( const create_proposal_operation& o )
+{
+   try
+   {
+      FC_ASSERT( _db.has_hardfork( STEEM_PROPOSALS_HARDFORK ), "Proposals functionality not enabled until hardfork ${hf}", ("hf", STEEM_PROPOSALS_HARDFORK) );
+
+      ilog("creating proposal: ${op}", ("op", o));
+
+      /** start date can be earlier than head_block_time - otherwise creating a proposal can be difficult,
+          since passed date should be adjusted by potential transaction execution delay (i.e. 3 sec
+          as a time for processed next block).
+      */
+      FC_ASSERT(o.end_date > _db.head_block_time(), "Can't create inactive proposals...");
+
+      asset fee_sbd( STEEM_TREASURY_FEE, SBD_SYMBOL );
+
+      FC_ASSERT( _db.get_balance( o.creator, SBD_SYMBOL ) >= fee_sbd,
+         "Account does not have sufficient funds for specified fee of ${of}", ("of", fee_sbd) );
+
+      //treasury account must exist, also we need it later to change its balance
+      const auto& treasury_account =_db.get_account( STEEM_TREASURY_ACCOUNT );
+
+      const auto& owner_account = _db.get_account( o.creator );
+      const auto* receiver_account = _db.find_account( o.receiver );
+
+      /// Just to check the receiver account exists.
+      FC_ASSERT(receiver_account != nullptr, "Specified receiver account: ${r} must exist in the blockchain",
+         ("r", o.receiver));
+
+      const auto* commentObject = _db.find_comment(o.creator, o.permlink);
+      if(commentObject == nullptr)
+      {
+         commentObject = _db.find_comment(o.receiver, o.permlink);
+         FC_ASSERT(commentObject != nullptr, "Proposal permlink must point to the article posted by creator or receiver");
+      }
+
+      _db.create< proposal_object >( [&]( proposal_object& proposal )
+      {
+         proposal.proposal_id = proposal.id;
+
+         proposal.creator = o.creator;
+         proposal.receiver = o.receiver;
+
+         proposal.start_date = o.start_date;
+         proposal.end_date = o.end_date;
+
+         proposal.daily_pay = o.daily_pay;
+
+         proposal.subject = o.subject.c_str();
+
+         proposal.permlink = o.permlink.c_str();
+      });
+
+      _db.adjust_balance( owner_account, -fee_sbd );
+      /// Fee shall be paid to the treasury
+      _db.adjust_balance(treasury_account, fee_sbd );
+   }
+   FC_CAPTURE_AND_RETHROW( (o) )
+}
+
+void update_proposal_votes_evaluator::do_apply( const update_proposal_votes_operation& o )
+{
+   try
+   {
+      FC_ASSERT( _db.has_hardfork( STEEM_PROPOSALS_HARDFORK ), "Proposals functionality not enabled until hardfork ${hf}", ("hf", STEEM_PROPOSALS_HARDFORK) );
+
+      ilog("voting proposal: ${op}", ("op", o));
+
+      if( o.proposal_ids.empty() )
```

Comment:
This is not needed. In `update_proposal_votes_operation::validate()` `proposal_ids.empty()` is asserted false.